pyro/ops/contract.py

```diff
@@ -487,3 +525,113 @@ def ubersum(equation, *operands, **kwargs):
+
+    :param dict dim_to_ordinal: a mapping from contraction dim to the set of
+        batch dims over which the contraction dim is batched.
+    """
+    def __init__(self, dim_to_ordinal):
+        self._plates = {d: tuple(sorted(ordinal)) for d, ordinal in dim_to_ordinal.items()}
+        self._symbols = map(opt_einsum.get_symbol, itertools.count())
+        self._map = {}
+
+    def __call__(self, dim, indices):
+        """
+        Converts a batched dim + batch indices to a flattened dim.
+
+        :param str dim: a batched dimension to flatten
+        :param dict indices: a mapping from batch dimension to int
+        :return: a flattened dim
+        :rtype: str
+        """
+        plate = self._plates.get(dim, ())
+        index = tuple(indices[d] for d in plate)
+        key = dim, index
+        if key in self._map:
+            return self._map[key]
+        normal_dim = next(self._symbols)
+        self._map[key] = normal_dim
+        return normal_dim
+
+
+def naive_ubersum(equation, *operands, **kwargs):
+    """
+    Naive reference implementation of :func:`ubersum`.
+
+    This implementation should never raise ``NotImplementedError``.
+    This implementation should agree with :func:`ubersum` whenver
+    :func:`ubersum` does not raise ``NotImplementedError``.
+    """
+    # Parse equation, without loss of generality assuming a single output.
+    inputs, outputs = equation.split('->')
+    outputs = outputs.split(',')
+    if len(outputs) > 1:
+        return tuple(naive_ubersum(inputs + '->' + output, *operands, **kwargs)[0]
+                     for output in outputs)
+    output, = outputs
+    inputs = inputs.split(',')
+
+    # Split dims into batch dims, contraction dims, and dims to keep.
+    batch_dims = set(kwargs.pop('batch_dims', ''))
+    if not batch_dims:
+        result = opt_einsum.contract(equation, *operands, backend='pyro.ops.einsum.torch_log')
+        return (result,)
+    output_dims = set(output)
+
+    # Collect sizes of all dimensions.
+    sizes = {}
+    for input_, operand in zip(inputs, operands):
+        for dim, size in zip(input_, operand.shape):
+            old = sizes.setdefault(dim, size)
+            if old != size:
+                raise ValueError(u"Dimension size mismatch at dim '{}': {} vs {}"
+                                 .format(dim, size, old))
+
+    # Compute batch context for each non-batch dim, by convention the
+    # intersection over all batch contexts of tensors in which the dim appears.
+    dim_to_ordinal = {}
+    for dims in map(set, inputs):
+        ordinal = dims & batch_dims
+        for dim in dims - batch_dims:
+            dim_to_ordinal[dim] = dim_to_ordinal.get(dim, ordinal) & ordinal
+    for dim in output_dims - batch_dims:
+        missing_dims = dim_to_ordinal[dim] - output_dims
+        if missing_dims:
+            raise ValueError(u"It is nonsensical to preserve a batched dim without preserving "
```

Comment:
Should this be preserving a "non-batched" dim?

Reply:
I'm trying to distinguish a dim-that-is-batched (i.e. a local enumeration dim) from a batch dim (i.e. an iarange dim). Martin suggested using the term `plate dim`. I guess we can hone that language in the paper then adjust the code.

Reply:
> I'm trying to distinguish a dim-that-is-batched (i.e. a local enumeration dim) from a batch dim (i.e. an iarange dim)

Makes sense. `plate dim` sounds like a reasonable choice.